dns_sd_bonjour.c

```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
 	CFStreamError* 			error,
 	void* 					info)
 {
-	CFStreamError anError;
-	
+	CFStreamError	anError;
+
 	if ((flags & kCFNetServiceFlagIsDomain) != 0) {
-		ERROR("DNS SD: FATAL. Callback called for domain, not service.\n");
+		ERROR("DNS SD: FATAL! Callback called for domain, not service.\n");
 		goto stop_browsing;
 	}
 	
-	struct cfnet_discovery_data *dd = (struct cfnet_discovery_data *)info;
+	struct dns_sd_discovery_data *dd = (struct dns_sd_discovery_data *)info;
 	if (dd == NULL) {
 		ERROR("DNS SD: Missing info structure. Stop browsing.\n");
 		goto stop_browsing;
 	}
-	
-	if (dd->have_v4 || dd->have_v6) {
-		// Already resolved one IIO service. Stop.
-		DEBUG("DNS SD: first service already resolved. Skip.\n");
-		goto stop_browsing;
+
+	if ((flags & kCFNetServiceFlagRemove) != 0) {
+		DEBUG("DNS SD: Callback to remove service. Ignore.\n");
+		return;
 	}
 
+	iio_mutex_lock(dd->lock);
+
 	const CFNetServiceRef netService = (CFNetServiceRef)domainOrService;
 	if (netService == NULL) {
-		ERROR("DNS SD: Net service is null.\n");
-		goto stop_browsing;
+		DEBUG("DNS SD: Net service is null.\n");
+		goto verify_flags;
 	}
 	
 	if (!CFNetServiceResolveWithTimeout(netService, 10.0, &anError)) {
-		ERROR("DNS SD: Resolve error: %ld.%d\n", anError.domain, anError.error);
-		goto stop_browsing;
+		DEBUG("DNS SD: Resolve error: %ld.%d\n", anError.domain, anError.error);
+		goto exit;
 	}
 	
-	dd->port = CFNetServiceGetPortNumber(netService);
+	CFStringRef targetHost = CFNetServiceGetTargetHost(netService);
+	if (targetHost == NULL) {
+		DEBUG("DNS SD: No valid target host for service.\n");
+		goto exit;
+	}
+
+	char hostname[MAXHOSTNAMELEN];
+	if (!CFStringGetCString(targetHost, hostname, sizeof(hostname), kCFStringEncodingASCII)) {
```

Comment:
Why is `kCFStringEncodingASCII` hardcoded here?

Would't it be better to user current system encoding?
```
CFStringEncoding encodingMethod = CFStringGetSystemEncoding();
```

Reply:
Not what I saw others doing.

The hostname, you need to pass to other networking functions, not print out to people, and the other system level networking functions want ASCII.

Reply:
OK.